doc/Precision_Guidelines_for_Ngen_Code_Validation.md

```diff
@@ -0,0 +1,36 @@
+
+
+**Experience to Date:  Comparison of Results of NGEN CFE codes with BMI and without BMI**
+
+In our tests of the Conceptual Functional Equivalent (CFE) versus CFE implemented via the Basic Model Interface (BMI) modularization methods, earlier comparisons checked six digits after the decimal points from the formatted outputs for catchment id-87. For these comparisons, we have seen identical values between the BMI-CFE and the original CFE code for Schaake Output Runoff, GIUH Runoff, Nash Lateral Flow, Ground Water Flow, and Total Discharge, for all 720 time steps, with 1 hour time step size. These tests are performed using varying initial conditions: soil_storage_percentage at 0.22, and 0.667; groundwater_storage_percentage at 0.5 and 0.01. We also looked at the total discharge data from standard screen output, which indicated 8-9 digits agreement between BMI-CFE and CFE calculations. In earlier tests, we have also shown similar agreement between CFE calculation results and the original T-Shirt C code calculation results.
+
+
+**Possible reasons for precision discrepancies between CFE and BMI-CFE:**
+
+The double precision numbers have an accuracy of 15 decimal digits, the loss of accuracy from rounding off error is expected, but precisely how many digit loss probably varies and depends on the specific code. Since BMI-CFE code has additional interface and difference in coding details, we hypothesize the memory allocation, cache usage, data sets loading into the registers may be different. These likely affect the rounding off processes and cause the least significant digits to differ. According to some literature, computer hardware architecture, compiler, even optimization may all affect numerical precision. 
+        
+In our CFE and BMI-CFE validation example, one particular case, deserves attention, which is: a key input variable, the rain rate, or specifically: precip_kg_per_m2, is a float point number, this could potentially reduces the precision in comparison between the two models even though all other variables are double, in consideration of rounding-off error in machine precision. There are many excellent discussions on rounding-off errors in literature due to machine precision, we refer interested readers to literature<sup>1-4</sup> for details. 

```

Comment:
One, `precip_kg_per_m2` is not _directly_ rain rate (or it would have a unit of time in the units?)

Two,  if the same float was provided to both implementations, are you  concluding that casting that number lost precision for the entire model which makes comparing outputs nessicary at the single precision level of accuracy (5 digits)?

Reply:
One: you are right, kg/m^2 translate into mm for water, according to hydrologist. I can delete the word rain rate. Thanks for the correction.

Two: Not entirely sure. If casting is completely random, we are left with  7 reliable digits. Our results seem to indicate we can achieve slightly better precision, 8 to 9 digits. Not sure exactly why.

Reply:
One thing to note in any code section where roundoff error is suspected is that the precision of floating point formats is much greater close to the value zero. This means that using normalized numbers can greatly reduce representation based error.

Reply:
precip_kg_per_m2, although declared as float in CFE code, it actually has many more than 7 significant digits and in scientific notation in the forcing file. But the exponent has 2 digits not 3 as in standard C++ double.